apps/network_fs/client/network_fs_client.cc

```diff
@@ -0,0 +1,540 @@
+#define FUSE_USE_VERSION 31
+
+#include "io.hh"
+
+#include <cassert>
+#include <cstring>
+#include <fuse.h>
+#include <iostream>
+#include <netdb.h>
+#include <optional>
+#include <sstream>
+#include <string>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <vector>
+
+using namespace std::string_literals;
+using std::vector;
+using std::string;
+using std::cerr;
+using std::cout;
+using std::endl;
+
+namespace {
+
+template<typename ...Args>
+void _debug(const char* func_name, int line, Args&&... args) {
+    std::stringstream s;
+    s << "\e[1;31mDEBUG \e[1;37m" << func_name << "()\e[0m \e[1;32m" << line << "\e[0m ";
+    (s << ... << args) << std::endl;
+    std::cerr << s.str();
+}
+
+#define DEBUG(args...) _debug(__FUNCTION__, __LINE__, ##args)
+
+// TODO: make options more c++ like (boost?)
+struct options {
+    const char *host;
+    const char *port;
+    int show_help;
+} options;
+
+#define OPTION(t, p) { t, offsetof(struct options, p), 1 }
+const struct fuse_opt option_spec[] = {
+    OPTION("--host=%s", host),
+    OPTION("--port=%s", port),
+    OPTION("-h", show_help),
+    OPTION("--help", show_help),
+    FUSE_OPT_END
+};
+
+void show_help(const char* progname) {
+    printf("usage: %s [options] <mountpoint>\n\n", progname);
+    printf("File-system specific options:\n"
+           "    --host=<s>          Host address to connect\n"
+           "                        (default: \"localhost\")\n"
+           "    --port=<s>          Port to use while connecting\n"
+           "                        (default: 6969)\n"
+           "\n");
+}
+
+template<typename F>
+class final_action {
+private:
+    F _clean;
+public:
+    final_action(F clean) : _clean(std::move(clean)) {}
+
+    final_action(const final_action&) = delete;
+    final_action(final_action&&) = delete;
+    final_action& operator=(const final_action&) = delete;
+    final_action& operator=(final_action&&) = delete;
+
+    ~final_action() {
+        try {
+            _clean();
+        } catch (...) {
+        }
+    }
+};
+
+template<typename F>
+final_action<F> finally(F f) {
```

Comment:
`seastar::defer` has pretty much identical semantics to `finally`, and `seastar::deferred_action` is an equivalent of `final_action` - let's just use them.